### PR TITLE
feat: Make it possible to pass any scikit-learn metric string to `report_metrics`

### LINKED ISSUE #1519: feat: Add all scikit-learn scores to report metrics string accepted
### Is your feature request related to a problem? Please describe.

I wanted to do the following code, but couldn't:  

```
datasets = fetch_employee_salaries()
X, y = datasets.X, datasets.y
X_train, X_test, y_train, y_test = train_test_split(X, y)
baseline_dummy = DummyRegressor()
baseline_dumb_est = EstimatorReport(
    baseline_dummy, X_train=X_train, y_train=y_train, X_test=X_test, y_test=y_test
)
baseline_dumb_cv.metrics.report_metrics(scoring=["neg_max_error"])
```

But I got the following (and clear) error:   
Image

### Describe the solution you'd like

I'd like to be able to give any pre-defined metric in scikit-learn:    
https://scikit-learn.org/1.5/modules/model_evaluation.html

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Muhammad-Rebaal**
Closes #1519 .

---

Hi @auguste-probabl , @MarieSacksick !

I've resolved the issue [#1519](https://github.com/probabl-ai/skore/issues/1519) in  the `report_metric` function to scores the report metrics string. Could you please review the PR .

Thank You !

### CODE REVIEW

**Comment by thomass-dev** [comment]
Thanks for your contribution. Please can you fix the failing tests and add a dedicated use-case in the unittests?

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py [hunk 1]
```diff
@@ -223,27 +232,104 @@ def report_metrics(
...
+                            "median_",
+                            "mutual_",
+                            "normalized_",
+                            "precision_",
+                            "r2_",
+                            "rand_",
+                            "recall_",
+                            "roc_auc_",
+                            "top_k_",
+                        )
```
Comment (auguste-probabl) [inline skore/src/skore/sklearn/_estimator/metrics_accessor.py:262]: Is this list really needed? Can't we just use `get_scorer` and catch the error if the metric does not exist?

Comment (Muhammad-Rebaal) [reply skore/src/skore/sklearn/_estimator/metrics_accessor.py:262]: Yes , we can do that . I'll implement that just in a while.

**Comment by Muhammad-Rebaal** [comment]
> Thanks for your contribution. Please can you fix the failing tests and add a dedicated use-case in the unittests?

Yes sure , I'll fix them and will add a use-case in the unittest.

**Comment by auguste-probabl** [comment]
> Just want to know that is there any test written for this function so that I'd modify that to pass the test ?

Tests would be in `tests/unit/sklearn/estimator/test_estimator.py`

**Comment by Muhammad-Rebaal** [comment]
Hi @thomass-dev  , @auguste-probabl  !

I've added the necessary test and fixed the failing tests . Could you please review my PR and let me know if there is anything else that needs to be done.

Thank You !

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py [hunk 2]
```diff
@@ -70,23 +70,19 @@ def report_metrics(
     ) -> pd.DataFrame:
         """Report a set of metrics for our estimator.
 
-        Parameters
+                Parameters
```
Comment (auguste-probabl) [inline skore/src/skore/sklearn/_estimator/metrics_accessor.py:73]: The indent shouldn't change
```suggestion
        Parameters
```

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py [hunk 3]
```diff
@@ -70,23 +70,19 @@ def report_metrics(
...
-
         X : array-like of shape (n_samples, n_features), default=None
             New data on which to compute the metric. By default, we use the validation
             set provided when creating the report.
-
         y : array-like of shape (n_samples,), default=None
             New target on which to compute the metric. By default, we use the target
             provided when creating the report.
-
         scoring : list of str, callable, or scorer, default=None
```
Comment (auguste-probabl) [inline skore/src/skore/sklearn/_estimator/metrics_accessor.py:86]: These vertical spaces help readability, please put them back

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py [hunk 4]
```diff
@@ -143,12 +134,13 @@ def report_metrics(
         Recall                 0.93...         (↗︎)
         ROC AUC                0.99...         (↗︎)
         Brier score            0.03...         (↘︎)
+        >>> # Using scikit-learn metrics
+        >>> report.metrics.report_metrics(scoring=["neg_log_loss"])
+                    LogisticRegression
+        Metric
+        Log Loss             -0.10...
```
Comment (auguste-probabl) [inline skore/src/skore/sklearn/_estimator/metrics_accessor.py:141]: It should rather say something like
```suggestion
        neg_log_loss             -0.10...
```
or maybe
```suggestion
        Negative Log Loss             -0.10...
```

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py [hunk 5]
```diff
@@ -143,12 +134,13 @@ def report_metrics(
...
         """
         if data_source == "X_y":
-            # optimization of the hash computation to avoid recomputing it
-            # FIXME: we are still recomputing the hash for all the metrics that we
-            # support in the report because we don't call `_compute_metric_scores`
-            # here. We should fix it.
```
Comment (auguste-probabl) [inline skore/src/skore/sklearn/_estimator/metrics_accessor.py:151]: Can you put the comments back, unless your code really demands changing them?

Comment (Muhammad-Rebaal) [reply skore/src/skore/sklearn/_estimator/metrics_accessor.py:151]: Thanks for reviewing, I'll put them back

**Comment by auguste-probabl** [comment]
Also noticed that your commits aren't signed. see https://docs.skore.probabl.ai/dev/contributing.html#signing-commits

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py [hunk 6]
```diff
@@ -134,11 +143,12 @@ def report_metrics(
         Recall                 0.93...         (↗︎)
         ROC AUC                0.99...         (↗︎)
         Brier score            0.03...         (↘︎)
+
         >>> # Using scikit-learn metrics
         >>> report.metrics.report_metrics(scoring=["neg_log_loss"])
                     LogisticRegression
         Metric
-        Log Loss             -0.10...
+        Negative Log Loss      -0.10...        (↘︎)
```
Comment (auguste-probabl) [inline skore/src/skore/sklearn/_estimator/metrics_accessor.py:151]: Need to add `Favorability` as a column

Comment (Muhammad-Rebaal) [reply skore/src/skore/sklearn/_estimator/metrics_accessor.py:151]: Doesn't understand your words can you clarify a little more

Comment (auguste-probabl) [reply skore/src/skore/sklearn/_estimator/metrics_accessor.py:151]: Yep sorry, please look at the previous example (the one just above) and add "Favorability" the same way

Comment (auguste-probabl) [reply skore/src/skore/sklearn/_estimator/metrics_accessor.py:151]: You should be able to test this using pytest

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 7]
```diff
@@ -1341,3 +1341,61 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+
+    report = EstimatorReport(
+        estimator, X_train=X_train, y_train=y_train, X_test=X_test, y_test=y_test
+    )
+
+    # Check that using brier_score with multiclass raises an error
+    with pytest.raises(
+        ValueError, match="Brier score is only defined for binary classification"
+    ):
+        report.metrics.brier_score()
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1401]: This test has nothing to do with the issue being solved; also, shouldn't it raise an `AttributeError`? Please split this out into another PR if you really want to have it merged

**Comment by auguste-probabl** [comment]
please sign your commits

**Comment by Muhammad-Rebaal** [comment]
> please sign your commits

Yeah , my last commit was signed.

**Comment by auguste-probabl** [comment]
> > please sign your commits
> 
> Yeah , my last commit was signed.

Awesome, thanks! However *all* your commits in the PR have to be signed

**Comment by Muhammad-Rebaal** [comment]
> > > please sign your commits
> > 
> > Yeah , my last commit was signed.
> 
> Awesome, thanks! However *all* your commits in the PR have to be signed

Yeah , it should be but besides all that learning signing commits is a new thing that I learned while contributing that's why the initial commits weren't signed. Although, I've added Favorability column and remove the test code which fixes the raising error , hopefully open a new issue for that . Is there anything else needed in this , let me know . I'd be happy to help.
Thank you

**Comment by thomass-dev** [comment]
Please, **all** your commits need to be signed.
One solution is to rebase your commits and squash all of it, before re-committing using `git commit -S`.

**Comment by Muhammad-Rebaal** [comment]
> Please, **all** your commits need to be signed. One solution is to rebase your commits and squash all of it, before re-committing using `git commit -S`.

Thanks @thomass-dev , I'll take a look at that and implement that .

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 8]
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+    from sklearn.datasets import load_breast_cancer
+    from sklearn.linear_model import LogisticRegression
+    from sklearn.metrics import get_scorer
+    from sklearn.model_selection import train_test_split
+
+    X_train, X_test, y_train, y_test = train_test_split(
+        *load_breast_cancer(return_X_y=True), random_state=0
+    )
+    classifier = LogisticRegression(max_iter=10_000)
+    classifier.fit(X_train, y_train)
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1361]:
```suggestion
```
EstimatorReport does the fitting

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 9]
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+        X_train=X_train,
+        y_train=y_train,
+        X_test=X_test,
+        y_test=y_test,
+    )
+
+    # Use scikit-learn's get_scorer to handle neg_log_loss
+    neg_log_loss_scorer = get_scorer("neg_log_loss")
+
+    result = report.metrics.report_metrics(scoring=[neg_log_loss_scorer])
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1374]:
```suggestion
    result = report.metrics.report_metrics(scoring=["neg_log_loss"])
```
The whole point of this change is to support passing scorers as strings.

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 10]
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+
+    # Check that the metric name is displayed properly (as 'log_loss')
+    assert "log_loss" in result.index
+
+    # Check that the value matches the absolute value of neg_log_loss
+    assert result.loc["log_loss", classifier.__class__.__name__] > 0
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1380]: That code doesn't do what the comment says.

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 11]
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
         result = getattr(report.metrics, metric_name)(average="macro")
         assert isinstance(result, float)
+
+
+def test_estimator_report_metric_with_neg_metrics():
+    """Check that scikit-learn metrics with 'neg_' prefix are handled correctly.
+
+    Non-regression test for:
+    https://github.com/probabl-ai/skore/issues/1519
+    """
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1351]:
```suggestion
    """Check that scikit-learn metrics with 'neg_' prefix are handled correctly."""
```

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 12]
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+    from sklearn.datasets import load_breast_cancer
+    from sklearn.linear_model import LogisticRegression
+    from sklearn.metrics import get_scorer
+    from sklearn.model_selection import train_test_split
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1355]: Move these outside

Comment (Muhammad-Rebaal) [reply skore/tests/unit/sklearn/estimator/test_estimator.py]: Yes, I've updated that and also remove redundancy.

File: skore/tests/unit/sklearn/estimator/test_estimator.py (hunk 10 above)
Comment (Muhammad-Rebaal) [reply skore/tests/unit/sklearn/estimator/test_estimator.py]: Yes , I've fixed that . It might be a mistake or so.

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py [hunk 13]
```diff
@@ -223,33 +221,79 @@ def report_metrics(
...
+                                "or a valid scikit-learn scoring string."
+                            ) from err
                     favorability_indicator.append(metric_favorability)
                 else:
+                    # Handle callable metrics
                     metric_fn = partial(self._custom_metric, metric_function=metric)
                     if scoring_kwargs is None:
                         metrics_kwargs = {}
                     else:
-                        # check if we should pass any parameters specific to the metric
```
Comment (auguste-probabl) [inline skore/src/skore/sklearn/_estimator/metrics_accessor.py:251]: put the comment back please

Comment (Muhammad-Rebaal) [reply skore/src/skore/sklearn/_estimator/metrics_accessor.py:251]: yeah sure

**Comment by Muhammad-Rebaal** [comment]
Hi @auguste-probabl !

Could you please review this PR and let me know if there is anything else needed to resolve ?

Thank You !

File: skore/src/skore/sklearn/_estimator/metrics_accessor.py (hunk 5 above)
Comment (auguste-probabl) [reply skore/src/skore/sklearn/_estimator/metrics_accessor.py:151]: this still needs to be done @Muhammad-Rebaal

File: skore/tests/unit/sklearn/estimator/test_estimator.py
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1]: Your PR makes it possible to write `report.metrics.report_metrics(scoring=["some_metric"])` (i.e. a string), so you have to test this

**Review by auguste-probabl — CHANGES_REQUESTED** [review]
Missing a test and some comments were removed which should not be

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 14]
```diff
@@ -1341,3 +1342,97 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+    classifier = LogisticRegression(max_iter=10_000)
+
+    report = EstimatorReport(
+        classifier,
+        X_train=X_train,
+        y_train=y_train,
+        X_test=X_test,
+        y_test=y_test,
+    )
+
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1362]: You should be able to use the `binary_classification_data` fixture:
```suggestion
def test_estimator_report_metric_with_neg_metrics(binary_classification_data):
    """Check that scikit-learn metrics with 'neg_' prefix are handled correctly."""

    classifier, X_test, y_test = binary_classification_data

    report = EstimatorReport(
        classifier,
        X_test=X_test,
        y_test=y_test,
    )
```
You might need to use a different metric than log loss since the estimator is a RandomForestClassifier.

Same for the other test.

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 15]
```diff
@@ -1341,3 +1342,97 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+        indicator_favorability=True,
+    )
+
+    assert "Mean Squared Error" in reg_result.index.get_level_values(0)
+    assert "Mean Absolute Error" in reg_result.index.get_level_values(0)
+    assert "R²" in reg_result.index.get_level_values(0)  # Changed from "R2" to "R²"
+
+    # Check favorability
+    assert reg_result.loc["Mean Squared Error"]["Favorability"] == "(↘︎)"
+    assert reg_result.loc["R²"]["Favorability"] == "(↗︎)"  # Changed from "R2" to "R²"
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1438]: Make a new test for this case.

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 16]
```diff
@@ -1341,3 +1342,92 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+    assert np.isclose(log_loss_value, abs(neg_log_loss_value))
+
+
+def test_estimator_report_with_sklearn_scoring_strings():
+    """Test that scikit-learn metric strings can be passed to report_metrics."""
+    # Classification test case
+    X_class, y_class = load_breast_cancer(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(
+        X_class, y_class, random_state=42
+    )
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1377]: Same as earlier, use a fixture

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 17]
```diff
@@ -1341,3 +1342,92 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+        X_reg, y_reg, random_state=42
+    )
+    regressor = LinearRegression()
+    reg_report = EstimatorReport(
+        regressor,
+        X_train=X_train_reg,
+        y_train=y_train_reg,
+        X_test=X_test_reg,
+        y_test=y_test_reg,
+    )
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1419]: Use a fixture

File: skore/tests/unit/sklearn/estimator/test_estimator.py [hunk 18]
```diff
@@ -1341,3 +1342,115 @@ def test_estimator_report_average_return_float(binary_classification_data):
...
+    X_train, X_test, y_train, y_test = train_test_split(
+        X_class, y_class, random_state=42
+    )
+    return (
+        LogisticRegression(max_iter=10_000).fit(X_train, y_train),
+        X_train,
+        X_test,
+        y_train,
+        y_test,
+    )
```
Comment (auguste-probabl) [inline skore/tests/unit/sklearn/estimator/test_estimator.py:1387]: There is already a fixture for binary classification, you just used it in the other test above. Same for regression, there is already a fixture you can use.
